### PR TITLE
Search fluent api concept intro

### PR DESCRIPTION
**Pull request by koltyakov**
| Q                       | A
| ----------------- | ---
| Bug fix?             | [ ]
| New feature?     | [x]
| New sample?     | [x]
| Related issues?  | mentioned in #207

#### What's in this Pull Request?

This PR introduces the concept which allows adding some fluent wrapper to search API, such as `startRow`, `rowLimit`, `pageNum` options for search queryable instance.

#### Guidance

`search` method in `Rest` class returns SearchResults promise, which prevents adding any additional options to a query in a fluent way.
An alternative approach is applied - `searchAdv` method is added to Rest class.
`searchAdv` returns Search object, allowing to define additional properties in a chain before actual execute.

#### Skipping results, defining results amount on a page

```javascript
var query = "*"; // or SearchQuery interface { Querytext: "*", ... }
$pnp.sp.searchAdv(query)
     .startRow(3) // An equivalent of providing StartRow property in SearchQuery options
     .rowLimit(7) // An equivalent of providing RowLimit property in SearchQuery options
     .execute().then(function(res) { 
           // Results of a search query processing
           // Will return 7 search results, 3 rows will be skipped
     });
```

#### Pagination helper

```javascript
var query = "*"; // or SearchQuery interface { Querytext: "*", ... }
$pnp.sp.searchAdv(query)
     // .rowLimit(15) // An equivalent of providing RowLimit property in SearchQuery options
     // When RowLimit is omited and pageNum is used a default value of 10 for RowLimit is used.
     // When pageNum is used StartRow setting declaration is ignored, value of StartRow is calculated automatically based on pageNum and RowLimit values. 
     .pageNum(2)
     .execute().then(function(res) { 
           // Results of a search query processing
           // Will return 10 search results, 10 rows will be skipped, showing a content for "2nd page"
     });
```

### CODE REVIEW

**Comment by koltyakov** [comment]
Hi Patrick (@patrick-rodgers),

There was an issue #207 sometime before, asking how to use pagination with search results.
I've decided to implement an alternative way to get pagination capability helper, which, I hope, is close to the vision of PnP-JS-Core. It's a working concept. I'm not totally sure about naming convention, but, in any case, I consider having an option defining some search request properties and calc some of them helpful.

Could you please take a look, is it ok with a whole structure of a library. Thanks!

**Comment by patrick-rodgers** [comment]
@koltyakov - sorry been slammed, will be looking at this hopefully this week but like what I see initially. I need to pull it down and really have a look - but didn't want you to think I'm ignoring the PR :)

File: src/sharepoint/rest/search.ts [hunk 1]
```diff
@@ -244,10 +244,19 @@ export class Search extends QueryableInstance {
...
-     * .......
+     *
      * @returns Promise
      */
-    public execute(query: SearchQuery): Promise<SearchResults> {
+    public execute(query?: SearchQuery): Promise<SearchResults> {
+
+        // Used with alternative search
+        if (typeof query === "undefined" && typeof this.searchQuery !== "undefined") {
+            if (typeof this.pageNumValue !== "undefined") {
```
Comment (mike-morrison) [inline src/sharepoint/rest/search.ts]: Do the defaults here match the OOB defaults when hitting the search endpoint without specifying values?

**Comment by johnnliu** [comment]
What about a slightly different syntax?

```
$pnp.sp.search(query)  // same as before
     .asRows()  // add this, also .asRows<type>()
     .startRow(3) // An equivalent of providing StartRow property in SearchQuery options
     .rowLimit(7) // An equivalent of providing RowLimit property in SearchQuery options
     .execute()
```

**Comment by patrick-rodgers** [comment]
Going to close this - but I took a lot of the ideas and used them in #433. Have a look and let me know what you think there. Will also be pushing these updates out in a beta shortly so folks can try it out and provide feedback.

**Comment by koltyakov** [comment]
Thanks Patrick, SearchQueryBuilder thing looks gracefully!

**Pull request closed without merging**